sakai-evaluation-impl/src/java/org/sakaiproject/evaluation/logic/externals/ExternalHierarchyLogicImpl.java

```diff
@@ -267,56 +497,152 @@ public void setEvalGroupsForNode(String nodeId, Set<String> evalGroupIds) {
+        }
+
+        for( String nodeId : nodeIds )
+        {
+            if( !m.containsKey( nodeId ) )
+            {
+                m.put( nodeId, 0 );
             }
         }
+
+        // Support for resolving eval groups based on hierarchy node rules
+        // Hierarchy rules should be obeyed regardless of if an external provider is present or not (supplemental)
+        Map<String, Set<String>> nodeGroupsMap = getEvalGroupsForNodes( nodeIds );
+        for( String nodeID : nodeGroupsMap.keySet() )
+        {
+            m.put( nodeID, nodeGroupsMap.get( nodeID ).size() );
+        }
+
         return m;
     }
 
     public List<EvalHierarchyNode> getNodesAboveEvalGroup(String evalGroupId) {
         List<EvalHierarchyNode> l = new ArrayList<EvalHierarchyNode>();
         if (evalHierarchyProvider != null) {
             l = evalHierarchyProvider.getNodesAboveEvalGroup(evalGroupId);
-        } else {
-            String nodeId = dao.getNodeIdForEvalGroup(evalGroupId);
-            if (nodeId != null) {
-                HierarchyNode currentNode = hierarchyService.getNodeById(nodeId);
-                Set<HierarchyNode> parents = hierarchyService.getParentNodes(nodeId, false);
-                parents.add(currentNode);
-                List<HierarchyNode> sorted = HierarchyUtils.getSortedNodes(parents);
-                // now convert the nodes to eval nodes
-                for (HierarchyNode node : sorted) {
-                    l.add( makeEvalNode(node) );
+        }
+
+        // Support for resolving eval groups based on hierarchy node rules
+        // Hierarchy rules should be obeyed regardless of if an external provider is present or not (supplemental)
+        String nodeID = "";
+        String siteID = evalGroupId.replace( EvalConstants.GROUP_ID_SITE_PREFIX, "" );
+        boolean isGroupIDSectionBased = false;
+        if( evalGroupId.contains( EvalConstants.GROUP_ID_SECTION_PREFIX ) )
+        {
+            siteID = siteID.substring( 0, siteID.indexOf( EvalConstants.GROUP_ID_SECTION_PREFIX ) );
+            isGroupIDSectionBased = true;
+        }
+
+        try
+        {
+            Site site = siteService.getSite( siteID );
+            List<HierarchyNodeRule> rules = externalLogic.getAllRules();
+            for( HierarchyNodeRule rule : rules )
+            {
+                String rawRuleText = removeQualifierFromRuleText( rule.getRule() );
+                String qualifier = determineQualifierFromRuleText( rule.getRule() );
+
+                // If the rule is section based and so is the group ID...
+                if( EvalConstants.HIERARCHY_RULE_SECTION.equals( rule.getOption() ) && isGroupIDSectionBased )
+                {
+                    // Get the sections of the site
+                    String realmID = siteService.siteReference( site.getId() );
+                    Set<String> sectionIDs = authzGroupService.getProviderIds( realmID );
```

Comment:
Does it make sense to store this result in case the next rule is also a section rule?